ICE: compiler/rustc_typeck/src/check/callee.rs:236:29: input to call/call_mut is not a ref?

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
#[lang = "fn"]
trait Fn {
    fn call(export_name);
}
fn call_through_fn_trait() {
    a()
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
binary: rustc
commit-hash: 07e0e2ec268c140e607e1ac7f49f145612d0f597
commit-date: 2021-03-24
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

### Error output

```
error[E0573]: expected type, found built-in attribute `export_name`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:13
  |
3 |     fn call(export_name);
  |             ^^^^^^^^^^^ not a type

error[E0425]: cannot find function `a` in this scope
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:6:5
  |
6 |     a()
  |     ^ not found in this scope

error[E0658]: language items are subject to change
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | #[lang = "fn"]
  | ^^^^^^^^^^^^^^
  |
  = help: add `#![feature(lang_items)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `perses_node_priority_with_dfs_delta_reduced_mutant`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | / #[lang = "fn"]
2 | | trait Fn {
3 | |     fn call(export_name);
4 | | }
5 | | fn call_through_fn_trait() {
6 | |     a()
7 | | }
  | |_^ consider adding a `main` function to `perses_node_priority_with_dfs_delta_reduced_mutant.rs`

error[E0152]: found duplicate lang item `fn`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:2:1
  |
2 | / trait Fn {
3 | |     fn call(export_name);
4 | | }
  | |_^
  |
  = note: the lang item is first defined in crate `core` (which `std` depends on)
  = note: first definition in `core` loaded from /home/cnsun/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-166dae07beec0398.rlib
  = note: second definition in the local crate (`perses_node_priority_with_dfs_delta_reduced_mutant`)

error: internal compiler error: compiler/rustc_typeck/src/check/callee.rs:236:29: input to call/call_mut is not a ref?
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:6:5
  |
6 |     a()
  |     ^^^

thread 'rustc' panicked at 'Box<Any>', /rustc/07e0e2ec268c140e607e1ac7f49f145612d0f597/library/std/src/panic.rs:59:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (07e0e2ec2 2021-03-24) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `call_through_fn_trait`
#1 [typeck_item_bodies] type-checking all item bodies
end of query stack
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0152, E0425, E0573, E0601, E0658.
For more information about an error, try `rustc --explain E0152`.
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0573]: expected type, found built-in attribute `export_name`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:13
  |
3 |     fn call(export_name);
  |             ^^^^^^^^^^^ not a type

error[E0425]: cannot find function `a` in this scope
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:6:5
  |
6 |     a()
  |     ^ not found in this scope

error[E0658]: language items are subject to change
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | #[lang = "fn"]
  | ^^^^^^^^^^^^^^
  |
  = help: add `#![feature(lang_items)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `perses_node_priority_with_dfs_delta_reduced_mutant`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | / #[lang = "fn"]
2 | | trait Fn {
3 | |     fn call(export_name);
4 | | }
5 | | fn call_through_fn_trait() {
6 | |     a()
7 | | }
  | |_^ consider adding a `main` function to `perses_node_priority_with_dfs_delta_reduced_mutant.rs`

error[E0152]: found duplicate lang item `fn`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:2:1
  |
2 | / trait Fn {
3 | |     fn call(export_name);
4 | | }
  | |_^
  |
  = note: the lang item is first defined in crate `core` (which `std` depends on)
  = note: first definition in `core` loaded from /home/cnsun/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-166dae07beec0398.rlib
  = note: second definition in the local crate (`perses_node_priority_with_dfs_delta_reduced_mutant`)

error: internal compiler error: compiler/rustc_typeck/src/check/callee.rs:236:29: input to call/call_mut is not a ref?
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:6:5
  |
6 |     a()
  |     ^^^

thread 'rustc' panicked at 'Box<Any>', /rustc/07e0e2ec268c140e607e1ac7f49f145612d0f597/library/std/src/panic.rs:59:5
stack backtrace:
   0: std::panicking::begin_panic
   1: std::panic::panic_any
   2: rustc_errors::HandlerInner::span_bug
   3: rustc_errors::Handler::span_bug
   4: rustc_middle::ty::context::tls::with_opt
   5: rustc_middle::util::bug::opt_span_bug_fmt
   6: rustc_middle::util::bug::span_bug_fmt
   7: rustc_typeck::check::callee::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::try_overloaded_call_traits
   8: rustc_typeck::check::callee::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_call
   9: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  10: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  11: rustc_typeck::check::fn_ctxt::checks::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected
  12: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  13: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_return_expr
  14: rustc_typeck::check::check::check_fn
  15: rustc_infer::infer::InferCtxtBuilder::enter
  16: rustc_typeck::check::typeck
  17: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  18: rustc_query_system::query::plumbing::force_query_with_job
  19: rustc_query_system::query::plumbing::get_query_impl
  20: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  21: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::par_body_owners
  22: rustc_typeck::check::typeck_item_bodies
  23: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  24: rustc_data_structures::stack::ensure_sufficient_stack
  25: rustc_query_system::query::plumbing::force_query_with_job
  26: rustc_query_system::query::plumbing::get_query_impl
  27: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck_item_bodies
  28: rustc_session::utils::<impl rustc_session::session::Session>::time
  29: rustc_typeck::check_crate
  30: rustc_interface::passes::analysis
  31: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  32: rustc_data_structures::stack::ensure_sufficient_stack
  33: rustc_query_system::query::plumbing::force_query_with_job
  34: rustc_query_system::query::plumbing::get_query_impl
  35: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  36: rustc_interface::passes::QueryContext::enter
  37: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  38: rustc_span::with_source_map
  39: rustc_interface::interface::create_compiler_and_run
  40: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (07e0e2ec2 2021-03-24) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `call_through_fn_trait`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0152, E0425, E0573, E0601, E0658.
For more information about an error, try `rustc --explain E0152`.
```

</p>
</details>

--- a/compiler/rustc_typeck/src/check/callee.rs
+++ b/compiler/rustc_typeck/src/check/callee.rs
@@ -246,12 +246,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 if borrow {
                     // Check for &self vs &mut self in the method signature. Since this is either
                     // the Fn or FnMut trait, it should be one of those.
-                    let (region, mutbl) =
-                        if let ty::Ref(r, _, mutbl) = method.sig.inputs()[0].kind() {
-                            (r, mutbl)
-                        } else {
-                            span_bug!(call_expr.span, "input to call/call_mut is not a ref?");
-                        };
+                    let (region, mutbl) = if let ty::Ref(r, _, mutbl) =
+                        method.sig.inputs()[0].kind()
+                    {
+                        (r, mutbl)
+                    } else {
+                        // The `fn`/`fn_mut` lang item is ill-formed, which should have
+                        // caused an error elsewhere.
+                        self.tcx
+                            .sess
+                            .delay_span_bug(call_expr.span, "input to call/call_mut is not a ref?");
+                        return None;
+                    };
 
                     let mutbl = match mutbl {
                         hir::Mutability::Not => AutoBorrowMutability::Not,

--- a/compiler/rustc_typeck/src/check/wfcheck.rs
+++ b/compiler/rustc_typeck/src/check/wfcheck.rs
@@ -203,25 +203,41 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
 
     let encl_trait_hir_id = tcx.hir().get_parent_item(hir_id);
     let encl_trait = tcx.hir().expect_item(encl_trait_hir_id);
-    if [tcx.lang_items().fn_trait(), tcx.lang_items().fn_mut_trait()]
-        .contains(&Some(encl_trait.def_id.to_def_id()))
-        && trait_item.ident.name.to_ident_string() == "call"
+    let encl_trait_def_id = encl_trait.def_id.to_def_id();
+    let fn_lang_item_name = if Some(encl_trait_def_id) == tcx.lang_items().fn_trait() {
+        Some("fn")
+    } else if Some(encl_trait_def_id) == tcx.lang_items().fn_mut_trait() {
+        Some("fn_mut")
+    } else {
+        None
+    };
+
+    if let (Some(fn_lang_item_name), "call") =
+        (fn_lang_item_name, trait_item.ident.name.to_ident_string().as_str())
     {
         // We are looking at the `call` function of the `fn` or `fn_mut` lang item.
         // Do some rudimentary sanity checking to avoid an ICE later (issue #83471).
-        if let Some(method_sig @ hir::FnSig { decl, .. }) = method_sig {
+        if let Some(hir::FnSig { decl, span, .. }) = method_sig {
             if let &[self_ty, _] = &decl.inputs {
                 if !matches!(self_ty.kind, hir::TyKind::Rptr(_, _)) {
-                    tcx.sess.struct_span_err(
-                        self_ty.span,
-                        "first argument of `call` in `fn`/`fn_mut` lang item must be a reference",
-                    ).emit();
+                    tcx.sess
+                        .struct_span_err(
+                            self_ty.span,
+                            &format!(
+                                "first argument of `call` in `{}` lang item must be a reference",
+                                fn_lang_item_name
+                            ),
+                        )
+                        .emit();
                 }
             } else {
                 tcx.sess
                     .struct_span_err(
-                        method_sig.span,
-                        "`call` function in `fn`/`fn_mut` lang item takes exactly two arguments",
+                        *span,
+                        &format!(
+                            "`call` function in `{}` lang item takes exactly two arguments",
+                            fn_lang_item_name
+                        ),
                     )
                     .emit();
             }
@@ -229,7 +245,10 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
             tcx.sess
                 .struct_span_err(
                     trait_item.span,
-                    "`call` trait item in `fn`/`fn_mut` lang item must be a function",
+                    &format!(
+                        "`call` trait item in `{}` lang item must be a function",
+                        fn_lang_item_name
+                    ),
                 )
                 .emit();
         }

--- a/compiler/rustc_typeck/src/check/wfcheck.rs
+++ b/compiler/rustc_typeck/src/check/wfcheck.rs
@@ -200,6 +200,40 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
     };
     check_object_unsafe_self_trait_by_name(tcx, &trait_item);
     check_associated_item(tcx, trait_item.hir_id(), span, method_sig);
+
+    let encl_trait_hir_id = tcx.hir().get_parent_item(hir_id);
+    let encl_trait = tcx.hir().expect_item(encl_trait_hir_id);
+    if [tcx.lang_items().fn_trait(), tcx.lang_items().fn_mut_trait()]
+        .contains(&Some(encl_trait.def_id.to_def_id()))
+        && trait_item.ident.name.to_ident_string() == "call"
+    {
+        // We are looking at the `call` function of the `fn` or `fn_mut` lang item.
+        // Do some rudimentary sanity checking to avoid an ICE later (issue #83471).
+        if let Some(method_sig @ hir::FnSig { decl, .. }) = method_sig {
+            if let &[self_ty, _] = &decl.inputs {
+                if !matches!(self_ty.kind, hir::TyKind::Rptr(_, _)) {
+                    tcx.sess.struct_span_err(
+                        self_ty.span,
+                        "first argument of `call` in `fn`/`fn_mut` lang item must be a reference",
+                    ).emit();
+                }
+            } else {
+                tcx.sess
+                    .struct_span_err(
+                        method_sig.span,
+                        "`call` function in `fn`/`fn_mut` lang item takes exactly two arguments",
+                    )
+                    .emit();
+            }
+        } else {
+            tcx.sess
+                .struct_span_err(
+                    trait_item.span,
+                    "`call` trait item in `fn`/`fn_mut` lang item must be a function",
+                )
+                .emit();
+        }
+    }
 }
 
 fn could_be_self(trait_def_id: LocalDefId, ty: &hir::Ty<'_>) -> bool {

--- a/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.rs
+++ b/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.rs
@@ -1,0 +1,27 @@
+// Make sure that an error is reported if the `call` function of the
+// `fn`/`fn_mut` lang item is grossly ill-formed.
+
+#![feature(lang_items)]
+#![feature(no_core)]
+#![no_core]
+
+#[lang = "fn"]
+trait MyFn<T> {
+    const call: i32 = 42;
+    //~^ ERROR: `call` trait item in `fn`/`fn_mut` lang item must be a function
+}
+
+#[lang = "fn_mut"]
+trait MyFnMut<T> {
+    fn call(i: i32, j: i32) -> i32 { i + j }
+    //~^ ERROR: first argument of `call` in `fn`/`fn_mut` lang item must be a reference
+}
+
+fn main() {
+    let a = || 42;
+    a();
+
+    let mut i = 0;
+    let mut b = || { i += 1; };
+    b();
+}

--- a/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.rs
+++ b/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.rs
@@ -8,13 +8,13 @@
 #[lang = "fn"]
 trait MyFn<T> {
     const call: i32 = 42;
-    //~^ ERROR: `call` trait item in `fn`/`fn_mut` lang item must be a function
+    //~^ ERROR: `call` trait item in `fn` lang item must be a function
 }
 
 #[lang = "fn_mut"]
 trait MyFnMut<T> {
     fn call(i: i32, j: i32) -> i32 { i + j }
-    //~^ ERROR: first argument of `call` in `fn`/`fn_mut` lang item must be a reference
+    //~^ ERROR: first argument of `call` in `fn_mut` lang item must be a reference
 }
 
 fn main() {

--- a/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.stderr
+++ b/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.stderr
@@ -1,0 +1,14 @@
+error: `call` trait item in `fn`/`fn_mut` lang item must be a function
+  --> $DIR/fn-fn_mut-call-ill-formed.rs:10:5
+   |
+LL |     const call: i32 = 42;
+   |     ^^^^^^^^^^^^^^^^^^^^^
+
+error: first argument of `call` in `fn`/`fn_mut` lang item must be a reference
+  --> $DIR/fn-fn_mut-call-ill-formed.rs:16:16
+   |
+LL |     fn call(i: i32, j: i32) -> i32 { i + j }
+   |                ^^^
+
+error: aborting due to 2 previous errors
+

--- a/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.stderr
+++ b/src/test/ui/lang-items/fn-fn_mut-call-ill-formed.stderr
@@ -1,10 +1,10 @@
-error: `call` trait item in `fn`/`fn_mut` lang item must be a function
+error: `call` trait item in `fn` lang item must be a function
   --> $DIR/fn-fn_mut-call-ill-formed.rs:10:5
    |
 LL |     const call: i32 = 42;
    |     ^^^^^^^^^^^^^^^^^^^^^
 
-error: first argument of `call` in `fn`/`fn_mut` lang item must be a reference
+error: first argument of `call` in `fn_mut` lang item must be a reference
   --> $DIR/fn-fn_mut-call-ill-formed.rs:16:16
    |
 LL |     fn call(i: i32, j: i32) -> i32 { i + j }

--- a/src/test/ui/lang-items/issue-83471.rs
+++ b/src/test/ui/lang-items/issue-83471.rs
@@ -1,0 +1,23 @@
+// Regression test for the ICE reported in issue #83471.
+
+#![crate_type="lib"]
+#![feature(no_core)]
+#![no_core]
+
+#[lang = "sized"]
+//~^ ERROR: language items are subject to change [E0658]
+trait Sized {}
+
+#[lang = "fn"]
+//~^ ERROR: language items are subject to change [E0658]
+//~| ERROR: `fn` language item must be applied to a trait with 1 generic argument
+trait Fn {
+    fn call(export_name);
+    //~^ ERROR: expected type
+    //~| WARNING: anonymous parameters are deprecated
+    //~| WARNING: this is accepted in the current edition
+}
+fn call_through_fn_trait() {
+    a()
+    //~^ ERROR: cannot find function
+}

--- a/src/test/ui/lang-items/issue-83471.stderr
+++ b/src/test/ui/lang-items/issue-83471.stderr
@@ -1,0 +1,51 @@
+error[E0573]: expected type, found built-in attribute `export_name`
+  --> $DIR/issue-83471.rs:15:13
+   |
+LL |     fn call(export_name);
+   |             ^^^^^^^^^^^ not a type
+
+error[E0425]: cannot find function `a` in this scope
+  --> $DIR/issue-83471.rs:21:5
+   |
+LL |     a()
+   |     ^ not found in this scope
+
+error[E0658]: language items are subject to change
+  --> $DIR/issue-83471.rs:7:1
+   |
+LL | #[lang = "sized"]
+   | ^^^^^^^^^^^^^^^^^
+   |
+   = help: add `#![feature(lang_items)]` to the crate attributes to enable
+
+error[E0658]: language items are subject to change
+  --> $DIR/issue-83471.rs:11:1
+   |
+LL | #[lang = "fn"]
+   | ^^^^^^^^^^^^^^
+   |
+   = help: add `#![feature(lang_items)]` to the crate attributes to enable
+
+warning: anonymous parameters are deprecated and will be removed in the next edition.
+  --> $DIR/issue-83471.rs:15:13
+   |
+LL |     fn call(export_name);
+   |             ^^^^^^^^^^^ help: try naming the parameter or explicitly ignoring it: `_: export_name`
+   |
+   = note: `#[warn(anonymous_parameters)]` on by default
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2018!
+   = note: for more information, see issue #41686 <https://github.com/rust-lang/rust/issues/41686>
+
+error[E0718]: `fn` language item must be applied to a trait with 1 generic argument
+  --> $DIR/issue-83471.rs:11:1
+   |
+LL | #[lang = "fn"]
+   | ^^^^^^^^^^^^^^
+...
+LL | trait Fn {
+   |         - this trait has 0 generic arguments
+
+error: aborting due to 5 previous errors; 1 warning emitted
+
+Some errors have detailed explanations: E0425, E0573, E0658, E0718.
+For more information about an error, try `rustc --explain E0425`.

